modules/internal/String.chpl

```diff
@@ -1756,402 +1761,404 @@ module String {
   }
 
   /*
-    Returns a new string, which is the concatenation of all of the strings
-    passed in with the receiving string inserted between them.
+    Returns a new :mod:`string <String>`, which is the concatenation of all of
+    the :mod:`string <String>` passed in with the contents of the method
+    receiver inserted between them.
 
     .. code-block:: chapel
 
         var myString = "|".join("a","10","d");
         writeln(myString); // prints: "a|10|d"
-   */
+
+    :arg x: :mod:`string <String>` values to be joined
+
+    :returns: A :mod:`string <String>`
+  */
   inline proc string.join(const ref x: string ...) : string {
-    return _join(x);
+    return doJoin(this, x);
   }
 
   /*
-    Same as the varargs version, but with a homogeneous tuple of strings.
+    Returns a new :mod:`string <String>`, which is the concatenation of all of
+    the :mod:`string <String>` passed in with the contents of the method
+    receiver inserted between them.
 
     .. code-block:: chapel
 
         var tup = ("a","10","d");
-        var myString = "|".join(tup);
-        writeln(myString); // prints: "a|10|d"
-   */
-  inline proc string.join(const ref x) : string where isTuple(x) {
-    if !isHomogeneousTuple(x) || !isString(x[1]) then
-      compilerError("join() on tuples only handles homogeneous tuples of strings");
-    return _join(x);
-  }
+        var myJoinedTuple = "|".join(tup);
+        writeln(myJoinedTuple); // prints: "a|10|d"
 
-  /*
-    Same as the varargs version, but with all the strings in an array.
+        var myJoinedArray = "|".join(["a","10","d"]);
+        writeln(myJoinedArray); // prints: "a|10|d"
 
-    .. code-block:: chapel
+    :arg x: An array or tuple of :mod:`string <String>` values to be joined
 
-        var myString = "|".join(["a","10","d"]);
-        writeln(myString); // prints: "a|10|d"
-   */
-  inline proc string.join(const ref x: [] string) : string {
-    return _join(x);
+    :returns: A :mod:`string <String>`
+  */
+  inline proc string.join(const ref x) : string {
+    compilerError("string.join() accepts any number of strings, homogenous "
```

Comment:
Whatever you choose to do for bytes.join (implement here vs add comment), please make the same change here, as well.